src/rpc.rs

```diff
@@ -12,80 +12,125 @@
  ********************************************************************************/
 use crate::UPClientZenoh;
 use async_trait::async_trait;
-use std::{string::ToString, time::Duration};
+use std::{string::ToString, sync::Arc, time::Duration};
 use tracing::error;
-use up_rust::{RpcClient, RpcClientResult, UAttributesError, UMessage, UMessageError, UUri};
+use up_rust::{
+    communication::{CallOptions, RpcClient, ServiceInvocationError, UPayload},
+    LocalUriProvider, UAttributes, UCode, UMessageType, UPayloadFormat, UPriority, UStatus, UUri,
+};
 use zenoh::prelude::r#async::*;
 
+pub struct ZenohRpcClient {
+    transport: Arc<UPClientZenoh>,
+    uri_provider: Arc<dyn LocalUriProvider>,
+}
+impl ZenohRpcClient {
+    /// Creates a new RPC client for the Zenoh transport.
+    ///
+    /// # Arguments
+    ///
+    /// * `transport` - The Zenoh uProtocol Transport Layer.
+    /// * `uri_provider` - The helper for creating URIs that represent local resources.
+    pub fn new(transport: Arc<UPClientZenoh>, uri_provider: Arc<dyn LocalUriProvider>) -> Self {
+        ZenohRpcClient {
+            transport,
+            uri_provider,
+        }
+    }
+}
+
 #[async_trait]
-impl RpcClient for UPClientZenoh {
-    // CY_TODO: Should remove method in the future
-    async fn invoke_method(&self, _method: UUri, request: UMessage) -> RpcClientResult {
-        // Get Zenoh key
-        let source = *request.attributes.source.0.clone().ok_or_else(|| {
-            let msg = "attributes.source should not be empty".to_string();
-            error!("{msg}");
-            UMessageError::PayloadError(msg)
-        })?;
-        let zenoh_key = if let Some(sink) = request.attributes.sink.0.clone() {
-            self.to_zenoh_key_string(&source, Some(&sink))
-        } else {
-            self.to_zenoh_key_string(&source, None)
+impl RpcClient for ZenohRpcClient {
+    async fn invoke_method(
+        &self,
+        method: UUri,
+        call_options: CallOptions,
+        payload: Option<UPayload>,
+    ) -> Result<Option<UPayload>, ServiceInvocationError> {
+        // Get source UUri
+        let source_uri = self.uri_provider.get_source_uri();
+
+        let attributes = UAttributes {
+            type_: UMessageType::UMESSAGE_TYPE_REQUEST.into(),
+            id: call_options.message_id().into(),
+            priority: call_options
+                .priority()
+                .unwrap_or(UPriority::UPRIORITY_UNSPECIFIED)
+                .into(),
+            source: Some(source_uri.clone()).into(),
+            sink: Some(method.clone()).into(),
+            ttl: Some(call_options.ttl()),
+            token: call_options.token(),
```

Comment:
you need to put the payload format into the attributes as well ...

Reply:
Yes, I forgot that...